command/login.go

```diff
@@ -157,37 +182,41 @@ func (c *LoginCommand) Run(args []string) int {
 		return 1
 	}
 
-	var token *oauth2.Token
-	switch {
-	case clientConfig.SupportedGrantTypes.Has(disco.OAuthAuthzCodeGrant):
-		// We prefer an OAuth code grant if the server supports it.
-		var tokenDiags tfdiags.Diagnostics
-		token, tokenDiags = c.interactiveGetTokenByCode(hostname, credsCtx, clientConfig)
-		diags = diags.Append(tokenDiags)
-		if tokenDiags.HasErrors() {
-			c.showDiagnostics(diags)
-			return 1
+	var token svcauth.HostCredentialsToken
+	var tokenDiags tfdiags.Diagnostics
+
+	// Prefer Terraform login if available
+	if clientConfig != nil {
```

Comment:
It looks like this `if` statement is redundant, since we've already exhausted the cases `if clientConfig == nil` - what do you think?

Reply:
I think it's still needed. The previous `if` statement doesn't result in an early exit, and only triggers an additional service discovery step. So we could still arrive at this line with a non-nil `clientConfig`, if the target host supports the `login.v1` service.

I don't think we can even move this block up to an `else` clause, because we want to print diagnostics and exit if neither service is supported. The control flow after these changes is definitely more confusing, but I can't find any way to simplify it.

Reply:
gotchya, that's really obvious in hindsight -  thanks!